src/main/scala/is/hail/keytable/KeyTable.scala

```diff
@@ -16,20 +16,90 @@ import org.apache.spark.SparkContext
 import org.apache.spark.rdd.RDD
 import org.apache.spark.sql.types.StructType
 import org.apache.spark.sql.{DataFrame, Row, SQLContext}
+import org.json4s.JObject
+import org.json4s.JsonAST._
+import org.json4s.jackson.{JsonMethods, Serialization}
 
 import scala.collection.JavaConverters._
 import scala.collection.mutable
 import scala.reflect.ClassTag
 
 
 object KeyTable {
+  final val fileVersion: Int = 1
+
   def fromDF(hc: HailContext, df: DataFrame, keyNames: Array[String]): KeyTable = {
     val signature = SparkAnnotationImpex.importType(df.schema).asInstanceOf[TStruct]
     KeyTable(hc, df.rdd.map { r =>
       SparkAnnotationImpex.importAnnotation(r, signature)
     },
       signature, keyNames)
   }
+
+  def read(hc: HailContext, path: String): KeyTable = {
+    if (!hc.hadoopConf.exists(path))
+      fatal(s"$path does not exist")
+
+    val metadataFile = path + "/metadata.json.gz"
+    val pqtSuccess = path + "/rdd.parquet/_SUCCESS"
+
+    if (!hc.hadoopConf.exists(pqtSuccess))
+      fatal(
+        s"corrupt KeyTable: parquet success file does not exist: $pqtSuccess")
+
+    if (!hc.hadoopConf.exists(metadataFile))
+      fatal(
+        s"corrupt KeyTable: metadata file does not exist: $metadataFile")
+
+    val (signature, keyNames) = try {
+      val json = hc.hadoopConf.readFile(metadataFile)(in =>
+        JsonMethods.parse(in))
+
+      val fields = json.asInstanceOf[JObject].obj.toMap
+
+      (fields.get("version"): @unchecked) match {
+        case Some(JInt(v)) =>
+          if (v != KeyTable.fileVersion)
+            fatal(
+              s"""Invalid KeyTable: old version
+                 |  got version $v, expected version ${ KeyTable.fileVersion }""".stripMargin)
+      }
+
+      val signature = (fields.get("schema"): @unchecked) match {
+        case Some(JString(s)) =>
+          Parser.parseType(s).asInstanceOf[TStruct]
+      }
+
+      val keyNames = (fields.get("key_names"): @unchecked) match {
+        case Some(JArray(a)) =>
+          a.map { case JString(s) => s
+          }.toArray[String]
+      }
+
+      (signature, keyNames)
+    } catch {
+      case e: Throwable =>
+        fatal(
+          s"""
+             |corrupt KeyTable: invalid metadata file.
+             |  caught exception: ${ expandException(e) }
+          """.stripMargin)
+    }
+
+    val requiresConversion = SparkAnnotationImpex.requiresConversion(signature)
+    val parquetFile = path + "/rdd.parquet"
+
+    val rdd = hc.sqlContext.read.parquet(parquetFile)
+      .rdd
+      .map { r =>
+        (if (requiresConversion)
+          SparkAnnotationImpex.importAnnotation(r, signature)
+        else
+          r): Any
```

Comment:
Mostly I wanted to keep the implementation of annotations abstract and not rely on the Spark stuff in case we want to change it (although we haven't done a great job of that).  I'd be OK for having a type alias for Row called `StructAnnotation` or something.  For another PR.